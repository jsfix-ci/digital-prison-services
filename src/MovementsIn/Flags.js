import React from 'react'
import PropTypes from 'prop-types'

import '../FullFlags/fullFlags.scss'
import OffenderLink from '../OffenderLink'

const Flags = ({ alerts, category, offenderNo }) => (
  <div className="full-flags">
    {alerts.includes('HA') && (
      <OffenderLink offenderNo={offenderNo} alerts>
        <span className="acct-status">ACCT OPEN</span>
      </OffenderLink>
    )}
    {alerts.includes('XSA') && (
      <OffenderLink offenderNo={offenderNo} alerts>
        <span className="assault-status">STAFF ASSAULTER</span>
      </OffenderLink>
    )}
    {alerts.includes('XA') && (
      <OffenderLink offenderNo={offenderNo} alerts>
        <span className="arsonist-status">
          <img src="/images/Arsonist_icon.png" className="arsonist-adjust" alt="" width="11" height="14" />
          <span> ARSONIST </span>
        </span>
      </OffenderLink>
    )}
    {alerts.includes('PEEP') && (
      <OffenderLink offenderNo={offenderNo} alerts>
        <span className="disability-status">
          <img src="/images/Disability_icon.png" className="disability-adjust" alt="" width="11" height="14" />
          <span> PEEP </span>
        </span>
      </OffenderLink>
    )}
    {alerts.includes('XEL') && (
      <OffenderLink offenderNo={offenderNo} alerts>
        <span className="elist-status">E&#x2011;LIST</span>
      </OffenderLink>
    )}
    {alerts.includes('XRF') && (
      <OffenderLink offenderNo={offenderNo} alerts>
        <span className="risk-females-status">RISK TO FEMALES</span>
      </OffenderLink>
    )}
    {alerts.includes('XTACT') && (
      <OffenderLink offenderNo={offenderNo} alerts>
        <span className="tact-status">TACT</span>
      </OffenderLink>
    )}
    {(category === 'A' || category === 'E') && <span className="cata-status">CAT&nbsp;A</span>}
    {category === 'H' && <span className="cata-high-status">CAT&nbsp;A&nbsp;High</span>}
    {category === 'P' && <span className="cata-prov-status">CAT&nbsp;A&nbsp;Prov</span>}
  </div>
)

Flags.propTypes = {
  alerts: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  category: PropTypes.string,
  offenderNo: PropTypes.string.isRequired,
}

Flags.defaultProps = {
  category: '',
}

export default Flags
