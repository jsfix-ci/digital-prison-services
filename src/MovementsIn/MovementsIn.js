import React from 'react'
import PropTypes from 'prop-types'
import SortableColumn from '../tablesorting/SortableColumn'
import OffenderName from '../OffenderName'
import DateFormatter from '../DateFormatter'
import OffenderLink from '../OffenderLink'
import OffenderImage from '../OffenderImage'
import SortLov from '../tablesorting/SortLov'
import Location from '../Location'
import HoursAndMinutes from '../HoursAndMinutes'
import { LAST_NAME } from '../tablesorting/sortColumns'
import Flags from '../Flags/Flags'
import FullFlags from '../FullFlags/FullFlags'

const movementInType = {
  offenderNo: PropTypes.string.isRequired,
  firstName: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  location: PropTypes.string.isRequired,
  movementTime: PropTypes.string,
  fromAgencyDescription: PropTypes.string,
  alerts: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  category: PropTypes.string,
}

const movementsInType = {
  movementsIn: PropTypes.arrayOf(PropTypes.shape(movementInType)).isRequired,
  sortOrder: PropTypes.string.isRequired,
  setColumnSort: PropTypes.func.isRequired,
}

const MovementsIn = props => {
  const { movementsIn, sortOrder, setColumnSort } = props
  return (
    <React.Fragment>
      <SortLov sortColumns={[LAST_NAME]} sortColumn={LAST_NAME} sortOrder={sortOrder} setColumnSort={setColumnSort} />
      <MovementsInTable movementsIn={movementsIn} sortOrder={sortOrder} setColumnSort={setColumnSort} />
    </React.Fragment>
  )
}

MovementsIn.propTypes = movementsInType

const MovementsInTable = ({ movementsIn, sortOrder, setColumnSort }) => (
  <table className="row-gutters">
    <MovementsInTableHeadings sortOrder={sortOrder} setColumnSort={setColumnSort} />
    <tbody>
      {movementsIn.map(movementIn => (
        <MovementInRow movementIn={movementIn} key={movementIn.offenderNo} />
      ))}
    </tbody>
  </table>
)

MovementsInTable.propTypes = movementsInType

const MovementsInTableHeadings = props => {
  const { sortOrder, setColumnSort } = props
  return (
    <thead>
      <tr>
        <th className="straight width10" />
        <th className="straight width10">
          <SortableColumn
            heading="Name"
            column={LAST_NAME}
            sortOrder={sortOrder}
            setColumnSort={setColumnSort}
            sortColumn={LAST_NAME}
          />
        </th>
        <th className="straight width10">Prison no.</th>
        <th className="straight width10">D.O.B.</th>
        <th className="straight width10">Location</th>
        <th className="straight width10">Time in</th>
        <th className="straight width15">Arrived from</th>
        <th className="straight width25">Flags</th>
      </tr>
    </thead>
  )
}
MovementsInTableHeadings.propTypes = {
  sortOrder: PropTypes.string.isRequired,
  setColumnSort: PropTypes.func.isRequired,
}

const MovementInRow = props => {
  const {
    movementIn: {
      alerts,
      category,
      dateOfBirth,
      firstName,
      fromAgencyDescription,
      lastName,
      location,
      movementTime,
      offenderNo,
    },
  } = props
  return (
    <tr className="row-gutters">
      <td className="row-gutters">
        <OffenderLink offenderNo={offenderNo}>
          <OffenderImage offenderNo={offenderNo} />
        </OffenderLink>
      </td>
      <td className="row-gutters">
        <OffenderLink offenderNo={offenderNo}>
          <OffenderName lastName={lastName} firstName={firstName} />
        </OffenderLink>
      </td>
      <td className="row-gutters">{offenderNo}</td>
      <td className="row-gutters">
        <DateFormatter isoDate={dateOfBirth} />
      </td>
      <td className="row-gutters">
        <Location location={location} />
      </td>
      <td>
        <HoursAndMinutes hhmmss={movementTime} />
      </td>
      <td>{fromAgencyDescription}</td>
      <td className="row-gutters">{FullFlags(alerts, category)}</td>
    </tr>
  )
}

MovementInRow.propTypes = {
  movementIn: PropTypes.shape(movementInType).isRequired,
}

export default MovementsIn
