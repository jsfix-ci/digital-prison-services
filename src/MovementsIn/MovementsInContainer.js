import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import axios from 'axios'

import MovementsIn from './MovementsIn'
import { resetError, setLoaded } from '../redux/actions'
import Error from '../Error'
import { ASC } from '../tablesorting/sortOrder'
import Page from '../Components/Page'
import routePaths from '../routePaths'

import SortableDataSource from '../tablesorting/SortableDataSource'
import { lastNameComparator } from '../tablesorting/comparatorComposition'

class MovementsInContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      movementsIn: [],
    }
  }

  componentDidMount() {
    this.getMovementsIn()
  }

  componentDidUpdate(previousProps) {
    const { agencyId, history } = this.props

    if (agencyId !== previousProps.agencyId) {
      history.replace(routePaths.establishmentRoll)
    }
  }

  async getMovementsIn() {
    const { agencyId, handleError, resetErrorDispatch, setLoadedDispatch } = this.props
    try {
      resetErrorDispatch()
      setLoadedDispatch(false)
      const response = await axios.get(`/api/movements/${agencyId}/in`)
      const movementsIn = response.data
      this.setState({ movementsIn })
      setLoadedDispatch(true)
    } catch (error) {
      handleError(error)
    }
  }

  render() {
    const { movementsIn, loaded } = this.state
    const { error } = this.props

    return (
      <Page title="In today" error={error} loaded={loaded}>
        <Error error={error} />
        <SortableDataSource sortOrder={ASC} rows={movementsIn} comparator={lastNameComparator}>
          <MovementsIn />
        </SortableDataSource>
      </Page>
    )
  }
}

MovementsInContainer.propTypes = {
  handleError: PropTypes.func.isRequired,
  history: PropTypes.shape({
    replace: PropTypes.func.isRequired,
  }).isRequired,

  // redux state
  agencyId: PropTypes.string.isRequired,
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.shape({ message: PropTypes.string })]),

  // redux dispatch
  resetErrorDispatch: PropTypes.func.isRequired,
}

MovementsInContainer.defaultProps = {
  error: '',
}

const mapStateToProps = state => ({
  agencyId: state.app.user.activeCaseLoadId,
  error: state.app.error,
})

const mapDispatchToProps = dispatch => ({
  resetErrorDispatch: () => dispatch(resetError()),
  setLoadedDispatch: status => dispatch(setLoaded(status)),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MovementsInContainer)
