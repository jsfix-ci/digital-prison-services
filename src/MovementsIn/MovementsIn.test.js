import React from 'react'
import { shallow } from 'enzyme'

import MovementsIn from './MovementsIn'

describe('Movements in', () => {
  it('should render page correctly', () => {
    const movements = [
      {
        offenderNo: 'G0000GG',
        firstName: 'firstName',
        lastName: 'lastName',
        location: 'MDI-1-3-017',
        movementTime: '12:01:02',
        fromAgencyDescription: 'Hull (HMP)',
        toAgencyId: 'MDI',
        alerts: ['HA', 'XSA', 'XA', 'PEEP', 'XEL', 'XRF', 'XTACT'],
      },
    ]
    const wrapper = shallow(<MovementsIn sortOrder="ASC" setColumnSort={jest.fn()} rows={movements} agencyId="MDI" />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should not make the offender clickable because there is no assigned location', () => {
    const movements = [
      {
        offenderNo: 'G0000GG',
        firstName: 'firstName',
        lastName: 'lastName',
        movementTime: '12:01:02',
        fromAgencyDescription: 'Hull (HMP)',
        toAgencyCode: 'HLI',
        alerts: ['HA', 'XSA', 'XA', 'PEEP', 'XEL', 'XRF', 'XTACT'],
      },
    ]
    const wrapper = shallow(<MovementsIn sortOrder="ASC" setColumnSort={jest.fn()} rows={movements} agencyId="HLI" />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should not make the offender clickable because there their toAgencyId does not match to active caseload', () => {
    const movements = [
      {
        offenderNo: 'G0000GG',
        firstName: 'firstName',
        lastName: 'lastName',
        movementTime: '12:01:02',
        location: 'CELL-1',
        fromAgencyDescription: 'Hull (HMP)',
        toAgencyCode: 'HLI',
        alerts: ['HA', 'XSA', 'XA', 'PEEP', 'XEL', 'XRF', 'XTACT'],
      },
    ]
    const wrapper = shallow(<MovementsIn sortOrder="ASC" setColumnSort={jest.fn()} rows={movements} agencyId="LEI" />)

    expect(wrapper).toMatchSnapshot()
  })
})
