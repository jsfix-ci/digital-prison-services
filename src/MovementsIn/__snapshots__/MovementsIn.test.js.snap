// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movements in should render page correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovementsIn
    movementsIn={
      Array [
        Object {
          "alerts": Array [
            "HA",
            "XSA",
            "XA",
            "PEEP",
            "XEL",
            "XRF",
            "XTACT",
          ],
          "firstName": "firstName",
          "fromAgencyDescription": "Hull (HMP)",
          "lastName": "lastName",
          "location": "MDI-1-3-017",
          "movementTime": "12:01:02",
          "offenderNo": "G0000GG",
        },
      ]
    }
    setColumnSort={[MockFunction]}
    sortOrder="ASC"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <SortLov
          setColumnSort={[MockFunction]}
          sortColumn="lastName"
          sortColumns={
            Array [
              "lastName",
            ]
          }
          sortOrder="ASC"
        />,
        <MovementsInTable
          movementsIn={
            Array [
              Object {
                "alerts": Array [
                  "HA",
                  "XSA",
                  "XA",
                  "PEEP",
                  "XEL",
                  "XRF",
                  "XTACT",
                ],
                "firstName": "firstName",
                "fromAgencyDescription": "Hull (HMP)",
                "lastName": "lastName",
                "location": "MDI-1-3-017",
                "movementTime": "12:01:02",
                "offenderNo": "G0000GG",
              },
            ]
          }
          setColumnSort={[MockFunction]}
          sortOrder="ASC"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "setColumnSort": [MockFunction],
          "sortColumn": "lastName",
          "sortColumns": Array [
            "lastName",
          ],
          "sortOrder": "ASC",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "movementsIn": Array [
            Object {
              "alerts": Array [
                "HA",
                "XSA",
                "XA",
                "PEEP",
                "XEL",
                "XRF",
                "XTACT",
              ],
              "firstName": "firstName",
              "fromAgencyDescription": "Hull (HMP)",
              "lastName": "lastName",
              "location": "MDI-1-3-017",
              "movementTime": "12:01:02",
              "offenderNo": "G0000GG",
            },
          ],
          "setColumnSort": [MockFunction],
          "sortOrder": "ASC",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <SortLov
            setColumnSort={[MockFunction]}
            sortColumn="lastName"
            sortColumns={
              Array [
                "lastName",
              ]
            }
            sortOrder="ASC"
          />,
          <MovementsInTable
            movementsIn={
              Array [
                Object {
                  "alerts": Array [
                    "HA",
                    "XSA",
                    "XA",
                    "PEEP",
                    "XEL",
                    "XRF",
                    "XTACT",
                  ],
                  "firstName": "firstName",
                  "fromAgencyDescription": "Hull (HMP)",
                  "lastName": "lastName",
                  "location": "MDI-1-3-017",
                  "movementTime": "12:01:02",
                  "offenderNo": "G0000GG",
                },
              ]
            }
            setColumnSort={[MockFunction]}
            sortOrder="ASC"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "setColumnSort": [MockFunction],
            "sortColumn": "lastName",
            "sortColumns": Array [
              "lastName",
            ],
            "sortOrder": "ASC",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "movementsIn": Array [
              Object {
                "alerts": Array [
                  "HA",
                  "XSA",
                  "XA",
                  "PEEP",
                  "XEL",
                  "XRF",
                  "XTACT",
                ],
                "firstName": "firstName",
                "fromAgencyDescription": "Hull (HMP)",
                "lastName": "lastName",
                "location": "MDI-1-3-017",
                "movementTime": "12:01:02",
                "offenderNo": "G0000GG",
              },
            ],
            "setColumnSort": [MockFunction],
            "sortOrder": "ASC",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
